# 06 — Consensus Backtest: Does Mutual-Fund Consensus Predict Returns?

**Question:** do stocks held by *more* mutual-fund schemes (`holders_total` in `consensus_panel`)
earn higher forward returns than stocks held by fewer schemes?

**Method (SQL-only against live Postgres — all business logic lives in the SQL statements below,
cells only orchestrate, count, and plot):**

1. For every `(isin, qtr)` row in `consensus_panel`, bucket the stock into a **decile of
   `holders_total` within its quarter** (`ntile(10) OVER (PARTITION BY qtr ORDER BY holders_total)`).
2. Compute the **t → t+4-quarter forward return per ISIN** from `security_prices`
   (entry = last close on/before quarter start, look-back tolerance 45 days;
   exit = close nearest to quarter start + 12 months, tolerance ±45 days).
3. Subtract the **NIFTY SMALLCAP 250** forward return over the *identical* window
   (same entry/exit date rules on `index_prices`) → **forward excess return**.
4. Report mean/median forward excess return by decile + per-quarter diagnostics + plot.

**Honesty guards (all counts printed below):**
- *Survivorship:* the panel is built from historical `portfolio_snapshots`, so it **includes exited
  holdings** (stocks later dropped by schemes still appear in their historical quarters) — this
  mitigates, but does not eliminate, survivorship bias (delisted symbols still have price gaps
  post-Jul-2024 per the ingestion caveats).
- *Missing prices:* panel rows whose ISIN has **no entry price** (mostly debt / repo / G-Sec / junk
  ISINs, which have no equity price history) are excluded and counted.
- *Short histories:* rows with an entry price but **no observable exit ≥ ~4 forward quarters**
  (no close within ±45 days of quarter start + 12 months) are excluded and counted.
- *Overlapping quarters:* panel coverage is deep only for recent quarters (B2 backfill got
  static-HTML AMCs), so usable (t, t+4q) windows are limited — every quarter's row count is printed.

**Compliance note:** this is a descriptive data-quality/sanity analysis of public disclosure data —
not investment advice.

In [1]:
import logging
import sys
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from sqlalchemy import create_engine, text

logging.basicConfig(level=logging.INFO, format="%(asctime)s | %(levelname)s | %(message)s",
                    stream=sys.stdout)
log = logging.getLogger("06_consensus_backtest")

REPO = Path.cwd()
while not (REPO / "AGENTS.md").exists() and REPO != REPO.parent:
    REPO = REPO.parent
REPORT_DIR = REPO / "data" / "reports" / "mutual_funds"
REPORT_DIR.mkdir(parents=True, exist_ok=True)

DB_URL = "postgresql+psycopg2://vlmrouter:vlmrouter@localhost:5432/mutual_funds"
eng = create_engine(DB_URL, pool_pre_ping=True)

def q(sql: str) -> pd.DataFrame:
    # run one SQL statement against live Postgres and return a DataFrame
    with eng.connect() as conn:
        return pd.read_sql(text(sql), conn)

BENCHMARK = "NIFTY SMALLCAP 250"
DEMO_ISIN, DEMO_NAME = "INE953O01021", "Sansera Engineering Limited"  # real stock held by small-cap schemes

with eng.connect() as conn:
    db_ok = conn.execute(text("SELECT version()")).scalar()
log.info("[STAGE 0] connected to Postgres: %s", db_ok.split(",")[0])
log.info("[STAGE 0] report dir: %s", REPORT_DIR)

2026-08-22 09:28:27,547 | INFO | [STAGE 0] connected to Postgres: PostgreSQL 18.3 (Debian 18.3-1.pgdg12+1) on aarch64-unknown-linux-gnu


2026-08-22 09:28:27,548 | INFO | [STAGE 0] report dir: /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds


In [2]:
log.info("[STAGE 1] consensus_panel overview (SQL view only)")

panel_by_qtr = q("""
    SELECT qtr,
           COUNT(*)                        AS n_rows,
           COUNT(DISTINCT isin)            AS n_isins,
           MIN(holders_total)              AS min_holders,
           MAX(holders_total)              AS max_holders,
           SUM((holders_total >= 2)::int)  AS rows_multi_holder
    FROM consensus_panel
    GROUP BY qtr
    ORDER BY qtr
""")
panel_tot = q("SELECT COUNT(*) AS n_rows, COUNT(DISTINCT isin) AS n_isins FROM consensus_panel")

print(f"consensus_panel total rows : {panel_tot.loc[0, 'n_rows']:,}")
print(f"consensus_panel unique ISINs: {panel_tot.loc[0, 'n_isins']:,}")
print(f"quarters with panel data   : {panel_by_qtr.shape[0]}")
print("\nRows per quarter:")
print(panel_by_qtr.to_string(index=False))

2026-08-22 09:28:27,551 | INFO | [STAGE 1] consensus_panel overview (SQL view only)


consensus_panel total rows : 18,251
consensus_panel unique ISINs: 7,730
quarters with panel data   : 21

Rows per quarter:
       qtr  n_rows  n_isins  min_holders  max_holders  rows_multi_holder
2017-07-01      28       28            1            1                  0
2021-04-01     850      850            1            8                312
2021-07-01    1034     1034            1           11                395
2021-10-01     146      146            1            1                  0
2022-01-01    1307     1307            1            6                320
2022-10-01     127      127            1            1                  0
2023-01-01      86       86            1            2                 30
2023-04-01     461      461            1           10                101
2023-07-01     845      845            1            3                 77
2023-10-01     696      696            1            6                 82
2024-01-01      30       30            1            1                  0
2

In [3]:
log.info("[STAGE 2] per-ISIN t -> t+4q forward returns from security_prices (SQL lateral joins)")

STOCK_FWD_SQL = """
WITH px AS (
    SELECT
        cp.isin,
        cp.qtr,
        cp.holders_total,
        e.close      AS entry_close,
        e.trade_date AS entry_date,
        x.close      AS exit_close,
        x.trade_date AS exit_date
    FROM consensus_panel cp
    -- entry: last close on/before quarter start (tolerance: 45 days look-back)
    LEFT JOIN LATERAL (
        SELECT sp.close, sp.trade_date
        FROM security_prices sp
        WHERE sp.isin = cp.isin
          AND sp.trade_date <= cp.qtr
          AND sp.trade_date >= cp.qtr - INTERVAL '45 days'
        ORDER BY sp.trade_date DESC
        LIMIT 1
    ) e ON TRUE
    -- exit: close nearest to quarter start + 12 months (tolerance: +/- 45 days)
    LEFT JOIN LATERAL (
        SELECT sp.close, sp.trade_date
        FROM security_prices sp
        WHERE sp.isin = cp.isin
          AND sp.trade_date >= cp.qtr + INTERVAL '12 months' - INTERVAL '45 days'
          AND sp.trade_date <= cp.qtr + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(sp.trade_date - (cp.qtr + INTERVAL '12 months')::date)
        LIMIT 1
    ) x ON TRUE
)
SELECT isin,
       qtr::date            AS qtr,
       holders_total,
       entry_date,
       entry_close::float8  AS entry_close,
       exit_date,
       exit_close::float8   AS exit_close
FROM px
"""

px = q(STOCK_FWD_SQL)
log.info("[STAGE 2] panel rows fetched: %d", len(px))

n_total = len(px)
no_entry = px["entry_close"].isna()
no_exit = px["entry_close"].notna() & px["exit_close"].isna()
ok = px["entry_close"].notna() & px["exit_close"].notna() & (px["entry_close"] > 0)

# classify the missing-entry-price rows: equity-looking ISINs vs debt/repo/gsec/junk
no_entry_eq = (no_entry & px["isin"].str.startswith("INE")).sum()
no_entry_other = (no_entry & ~px["isin"].str.startswith("INE")).sum()

print(f"[STAGE 2] panel rows (t observations)                     : {n_total:,}")
print(f"[STAGE 2] excluded - no entry price (missing price)       : {int(no_entry.sum()):,}"
      f"  (equity-looking INE ISINs: {int(no_entry_eq):,}; debt/repo/gsec/junk ISINs: {int(no_entry_other):,})")
print(f"[STAGE 2] excluded - entry ok but no ~4q forward price    : {int(no_exit.sum()):,}"
      f"  (short forward history / delisted gap)")
print(f"[STAGE 2] included - full entry + ~4q forward window      : {int(ok.sum()):,}")
assert int(no_entry.sum() + no_exit.sum() + ok.sum()) == n_total

stock_fwd = px.loc[ok, ["isin", "qtr", "holders_total", "entry_date", "entry_close",
                        "exit_date", "exit_close"]].copy()
stock_fwd["fwd_ret"] = stock_fwd["exit_close"] / stock_fwd["entry_close"] - 1.0
print("\n[STAGE 2] stock forward-return observations by quarter:")
print(stock_fwd.groupby("qtr").agg(n_obs=("isin", "size"), n_isins=("isin", "nunique")).to_string())

2026-08-22 09:28:27,603 | INFO | [STAGE 2] per-ISIN t -> t+4q forward returns from security_prices (SQL lateral joins)


2026-08-22 09:28:29,844 | INFO | [STAGE 2] panel rows fetched: 18251


[STAGE 2] panel rows (t observations)                     : 18,251
[STAGE 2] excluded - no entry price (missing price)       : 14,846  (equity-looking INE ISINs: 9,744; debt/repo/gsec/junk ISINs: 5,102)
[STAGE 2] excluded - entry ok but no ~4q forward price    : 1,770  (short forward history / delisted gap)
[STAGE 2] included - full entry + ~4q forward window      : 1,635

[STAGE 2] stock forward-return observations by quarter:
            n_obs  n_isins
qtr                       
2017-07-01     17       17
2021-04-01     58       58
2021-07-01     61       61
2021-10-01    138      138
2022-01-01    133      133
2022-10-01      6        6
2023-01-01     82       82
2023-04-01     21       21
2023-07-01    329      329
2023-10-01     65       65
2024-04-01    352      352
2024-07-01     50       50
2024-10-01    138      138
2025-01-01     27       27
2025-04-01     67       67
2025-10-01     91       91


In [4]:
log.info("[STAGE 3] NIFTY SMALLCAP 250 forward return over identical windows (SQL)")

IDX_FWD_SQL = f"""
WITH qtrs AS (
    SELECT DISTINCT qtr FROM consensus_panel
)
SELECT q.qtr::date AS qtr,
       e.close::float8      AS entry_close,
       e.trade_date         AS entry_date,
       x.close::float8      AS exit_close,
       x.trade_date         AS exit_date,
       (x.close / e.close - 1)::float8 AS fwd_ret
FROM qtrs q
JOIN LATERAL (
    SELECT ip.close, ip.trade_date
    FROM index_prices ip
    WHERE ip.index_symbol = '{BENCHMARK}'
      AND ip.trade_date <= q.qtr
    ORDER BY ip.trade_date DESC
    LIMIT 1
) e ON TRUE
JOIN LATERAL (
    SELECT ip.close, ip.trade_date
    FROM index_prices ip
    WHERE ip.index_symbol = '{BENCHMARK}'
      AND ip.trade_date >= q.qtr + INTERVAL '12 months' - INTERVAL '45 days'
      AND ip.trade_date <= q.qtr + INTERVAL '12 months' + INTERVAL '45 days'
    ORDER BY ABS(ip.trade_date - (q.qtr + INTERVAL '12 months')::date)
    LIMIT 1
) x ON TRUE
"""

idx_fwd = q(IDX_FWD_SQL)
log.info("[STAGE 3] %s windows computed for %d quarters", BENCHMARK, len(idx_fwd))

n_qtrs_panel = panel_by_qtr.shape[0]
n_qtrs_idx = idx_fwd["qtr"].nunique()
print(f"[STAGE 3] panel quarters                                  : {n_qtrs_panel}")
print(f"[STAGE 3] quarters with a full benchmark 4q window        : {n_qtrs_idx}")
print(f"[STAGE 3] quarters dropped (benchmark window incomplete)  : {n_qtrs_panel - n_qtrs_idx}")
print(f"\n[STAGE 3] {BENCHMARK} forward 12m return by quarter:")
print(idx_fwd.sort_values("qtr").to_string(index=False))

2026-08-22 09:28:29,855 | INFO | [STAGE 3] NIFTY SMALLCAP 250 forward return over identical windows (SQL)


2026-08-22 09:28:29,866 | INFO | [STAGE 3] NIFTY SMALLCAP 250 windows computed for 18 quarters


[STAGE 3] panel quarters                                  : 21
[STAGE 3] quarters with a full benchmark 4q window        : 18
[STAGE 3] quarters dropped (benchmark window incomplete)  : 3

[STAGE 3] NIFTY SMALLCAP 250 forward 12m return by quarter:
       qtr  entry_close entry_date  exit_close  exit_date   fwd_ret
2017-07-01      5982.50 2017-06-30     5733.78 2018-07-02 -0.041575
2021-04-01      7089.20 2021-04-01     9620.15 2022-04-01  0.357015
2021-07-01      8502.10 2021-07-01     8105.50 2022-07-01 -0.046647
2021-10-01      9391.60 2021-10-01     9198.75 2022-09-30 -0.020534
2022-01-01      9840.00 2021-12-31     9552.80 2023-01-02 -0.029187
2022-10-01      9198.75 2022-09-30    12230.40 2023-09-29  0.329572
2023-01-01      9481.25 2022-12-30    14129.70 2024-01-01  0.490278
2023-04-01      8787.90 2023-03-31    14762.15 2024-04-01  0.679827
2023-07-01     10544.20 2023-06-30    17367.65 2024-07-01  0.647128
2023-10-01     12230.40 2023-09-29    18531.80 2024-10-01  0.515224
202

In [5]:
log.info("[STAGE 4] decile bucketing (ntile within quarter) + excess return, aggregated in SQL")

DECILE_SQL = """
WITH px AS (
    -- identical forward-return SQL as STAGE 2, inlined so the whole pipeline is one statement
    SELECT cp.isin, cp.qtr, cp.holders_total,
           e.close AS entry_close, x.close AS exit_close
    FROM consensus_panel cp
    LEFT JOIN LATERAL (
        SELECT sp.close FROM security_prices sp
        WHERE sp.isin = cp.isin AND sp.trade_date <= cp.qtr
          AND sp.trade_date >= cp.qtr - INTERVAL '45 days'
        ORDER BY sp.trade_date DESC LIMIT 1
    ) e ON TRUE
    LEFT JOIN LATERAL (
        SELECT sp.close FROM security_prices sp
        WHERE sp.isin = cp.isin
          AND sp.trade_date >= cp.qtr + INTERVAL '12 months' - INTERVAL '45 days'
          AND sp.trade_date <= cp.qtr + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(sp.trade_date - (cp.qtr + INTERVAL '12 months')::date) LIMIT 1
    ) x ON TRUE
),
stock_fwd AS (
    SELECT isin, qtr, holders_total, (exit_close / entry_close - 1)::float8 AS fwd_ret
    FROM px WHERE entry_close > 0 AND exit_close IS NOT NULL
),
idx_fwd AS (
    SELECT q.qtr, (x.close / e.close - 1)::float8 AS fwd_ret
    FROM (SELECT DISTINCT qtr FROM stock_fwd) q
    JOIN LATERAL (
        SELECT ip.close FROM index_prices ip
        WHERE ip.index_symbol = '__BENCH__' AND ip.trade_date <= q.qtr
        ORDER BY ip.trade_date DESC LIMIT 1
    ) e ON TRUE
    JOIN LATERAL (
        SELECT ip.close FROM index_prices ip
        WHERE ip.index_symbol = '__BENCH__'
          AND ip.trade_date >= q.qtr + INTERVAL '12 months' - INTERVAL '45 days'
          AND ip.trade_date <= q.qtr + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(ip.trade_date - (q.qtr + INTERVAL '12 months')::date) LIMIT 1
    ) x ON TRUE
),
scored AS (
    SELECT s.isin, s.qtr, s.holders_total, s.fwd_ret, i.fwd_ret AS idx_fwd_ret,
           s.fwd_ret - i.fwd_ret AS excess_ret
    FROM stock_fwd s JOIN idx_fwd i USING (qtr)
),
deciled AS (
    SELECT *, ntile(10) OVER (PARTITION BY qtr ORDER BY holders_total, isin) AS decile
    FROM scored
)
SELECT decile,
       COUNT(*)                                                      AS n_obs,
       COUNT(DISTINCT isin)                                          AS n_isins,
       COUNT(DISTINCT qtr)                                           AS n_qtrs,
       AVG(holders_total)                                            AS avg_holders,
       AVG(excess_ret) * 100                                         AS mean_excess_pct,
       percentile_cont(0.5) WITHIN GROUP (ORDER BY excess_ret) * 100 AS median_excess_pct,
       stddev_samp(excess_ret) * 100                                 AS sd_excess_pct
FROM deciled
GROUP BY decile
ORDER BY decile
""".replace("__BENCH__", BENCHMARK)

decile_tbl = q(DECILE_SQL)
decile_tbl["mean_excess_pct"] = decile_tbl["mean_excess_pct"].astype(float).round(3)
decile_tbl["median_excess_pct"] = decile_tbl["median_excess_pct"].astype(float).round(3)
decile_tbl["sd_excess_pct"] = decile_tbl["sd_excess_pct"].astype(float).round(3)
decile_tbl["avg_holders"] = decile_tbl["avg_holders"].astype(float).round(1)

print(f"[STAGE 4] decile bucket sizes (n_obs) — total {int(decile_tbl['n_obs'].sum()):,} observations:")
print(decile_tbl[["decile", "n_obs", "n_isins", "n_qtrs", "avg_holders"]].to_string(index=False))
print("\n[STAGE 4] mean / median forward excess return vs NIFTY SMALLCAP 250 by holders_total decile (%):")
print(decile_tbl[["decile", "n_obs", "mean_excess_pct", "median_excess_pct", "sd_excess_pct"]]
      .to_string(index=False))

2026-08-22 09:28:29,872 | INFO | [STAGE 4] decile bucketing (ntile within quarter) + excess return, aggregated in SQL


[STAGE 4] decile bucket sizes (n_obs) — total 1,635 observations:
 decile  n_obs  n_isins  n_qtrs  avg_holders
      1    171       84      16          1.0
      2    168      106      16          1.0
      3    166      117      16          1.0
      4    165      115      16          1.2
      5    165      130      16          1.3
      6    164      128      16          1.6
      7    163      120      15          1.8
      8    160      119      15          2.2
      9    157      110      15          3.0
     10    156       80      15          4.8

[STAGE 4] mean / median forward excess return vs NIFTY SMALLCAP 250 by holders_total decile (%):
 decile  n_obs  mean_excess_pct  median_excess_pct  sd_excess_pct
      1    171           -9.183            -10.664         38.380
      2    168           -8.610             -9.971         33.296
      3    166           -3.290             -6.744         38.384
      4    165            0.673             -5.442         42.263
      5    

In [6]:
log.info("[STAGE 5] per-quarter diagnostics + monotonicity checks")

# raw scored+deciled rows (same pipeline, unaggregated) for per-quarter stats
_RAW_AGG = """
SELECT decile,
       COUNT(*)                                                      AS n_obs,
       COUNT(DISTINCT isin)                                          AS n_isins,
       COUNT(DISTINCT qtr)                                           AS n_qtrs,
       AVG(holders_total)                                            AS avg_holders,
       AVG(excess_ret) * 100                                         AS mean_excess_pct,
       percentile_cont(0.5) WITHIN GROUP (ORDER BY excess_ret) * 100 AS median_excess_pct,
       stddev_samp(excess_ret) * 100                                 AS sd_excess_pct
FROM deciled
GROUP BY decile
ORDER BY decile
"""
RAW_SQL = DECILE_SQL.replace(_RAW_AGG,
    "SELECT isin, qtr::date AS qtr, holders_total, fwd_ret, idx_fwd_ret, excess_ret, decile FROM deciled")
raw = q(RAW_SQL)
log.info("[STAGE 5] raw scored observations: %d across %d quarters", len(raw), raw["qtr"].nunique())

per_qtr = raw.groupby("qtr").apply(
    lambda g: pd.Series({
        "n_obs": len(g),
        "d1_mean_excess_pct": g.loc[g.decile == 1, "excess_ret"].mean() * 100,
        "d10_mean_excess_pct": g.loc[g.decile == 10, "excess_ret"].mean() * 100,
        "d10_minus_d1_pct": (g.loc[g.decile == 10, "excess_ret"].mean()
                             - g.loc[g.decile == 1, "excess_ret"].mean()) * 100,
        "spearman_dec_vs_excess": g[["decile", "excess_ret"]].corr(method="spearman").iloc[0, 1],
    }),
    include_groups=False,
)
print("[STAGE 5] per-quarter decile signal (mean excess %, D10-D1 spread, within-quarter Spearman):")
print(per_qtr.round(3).to_string())

# pooled monotonicity of the mean-excess decile profile
means = decile_tbl["mean_excess_pct"].to_numpy()
adj_up = int(np.sum(np.diff(means) > 0))
adj_down = int(np.sum(np.diff(means) < 0))
strictly_monotonic = bool(np.all(np.diff(means) > 0)) or bool(np.all(np.diff(means) < 0))
pearson_all = raw[["decile", "excess_ret"]].corr(method="pearson").iloc[0, 1]
spearman_all = raw[["decile", "excess_ret"]].corr(method="spearman").iloc[0, 1]

print(f"\n[STAGE 5] adjacent-decile mean-excess steps: UP {adj_up}/9, DOWN {adj_down}/9")
print(f"[STAGE 5] pooled mean-excess profile strictly monotonic? {strictly_monotonic}")
print(f"[STAGE 5] pooled Pearson  corr(decile, excess_ret): {pearson_all:+.4f}")
print(f"[STAGE 5] pooled Spearman corr(decile, excess_ret): {spearman_all:+.4f}")
print(f"[STAGE 5] quarters with positive D10-D1 spread: "
      f"{int((per_qtr['d10_minus_d1_pct'] > 0).sum())} / {len(per_qtr)}")

2026-08-22 09:28:30,964 | INFO | [STAGE 5] per-quarter diagnostics + monotonicity checks


2026-08-22 09:28:31,971 | INFO | [STAGE 5] raw scored observations: 1635 across 16 quarters


[STAGE 5] per-quarter decile signal (mean excess %, D10-D1 spread, within-quarter Spearman):
            n_obs  d1_mean_excess_pct  d10_mean_excess_pct  d10_minus_d1_pct  spearman_dec_vs_excess
qtr                                                                                                 
2017-07-01   17.0               5.914              -44.458           -50.373                   0.030
2021-04-01   58.0              -7.823               -4.257             3.566                   0.046
2021-07-01   61.0              -1.620                3.842             5.462                   0.117
2021-10-01  138.0             -11.254              -21.964           -10.709                  -0.106
2022-01-01  133.0              -3.047                3.318             6.365                  -0.069
2022-10-01    6.0             227.278                  NaN               NaN                  -0.257
2023-01-01   82.0             -15.747              -19.814            -4.066                  -0.20

In [7]:
log.info("[STAGE 6] plot + artifacts")

png_path = REPORT_DIR / "06_consensus_decile_backtest.png"
csv_path = REPORT_DIR / "06_consensus_decile_backtest.csv"

decile_tbl.to_csv(csv_path, index=False)
log.info("[STAGE 6] wrote %s", csv_path)

fig, ax = plt.subplots(figsize=(10, 5.5))
x = decile_tbl["decile"].to_numpy()
w = 0.38
ax.bar(x - w / 2, decile_tbl["mean_excess_pct"], width=w, label="mean", color="#3b6ea5")
ax.bar(x + w / 2, decile_tbl["median_excess_pct"], width=w, label="median", color="#c46a1f")
ax.axhline(0, color="black", linewidth=0.8)
ax.set_xticks(x)
ax.set_xlabel("holders_total decile within quarter (1 = fewest schemes, 10 = most schemes)")
ax.set_ylabel("forward 12m excess return vs NIFTY SMALLCAP 250 (%)")
ax.set_title("Mutual-fund consensus (holders_total) decile vs forward excess return\n"
             f"consensus_panel x security_prices, benchmark {BENCHMARK}")
ax.legend()
fig.tight_layout()
fig.savefig(png_path, dpi=150)
plt.close(fig)
log.info("[STAGE 6] wrote %s", png_path)

print(f"[STAGE 6] plot saved : {png_path}")
print(f"[STAGE 6] table saved: {csv_path}")

2026-08-22 09:28:31,989 | INFO | [STAGE 6] plot + artifacts


2026-08-22 09:28:31,991 | INFO | [STAGE 6] wrote /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds/06_consensus_decile_backtest.csv


2026-08-22 09:28:32,098 | INFO | [STAGE 6] wrote /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds/06_consensus_decile_backtest.png


[STAGE 6] plot saved : /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds/06_consensus_decile_backtest.png
[STAGE 6] table saved: /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds/06_consensus_decile_backtest.csv


## Honest Conclusion — does consensus predict returns?

**No monotonic decile relationship is present in this data.** The pipeline runs end-to-end on live
Postgres, but the "signal" is essentially null:

- **Coverage:** 18,251 panel rows across 21 quarters → **14,846** excluded (no entry price:
  5,102 debt/repo/G-Sec/junk ISINs + 9,744 equity-looking ISINs without price coverage),
  **1,770** excluded (entry ok but no ~4-quarter observable forward window / delisted gap),
  leaving **1,635** scored observations across **16 usable quarters**.
- **Decile profile of mean forward excess return vs NIFTY SMALLCAP 250 (%):**
  D1 −9.18, D2 −8.61, D3 −3.29, D4 **+0.67**, D5 −0.65, D6 −5.44, D7 −7.34, D8 −6.65,
  D9 −6.15, D10 −7.77. Median profile is negative everywhere except a shallow hump around D3–D5.
- **Monotonicity:** adjacent-decile mean-excess steps go UP 5/9 and DOWN 4/9 — not monotonic.
  Pooled Pearson corr(decile, excess) = **−0.006**, Spearman = **−0.004**: statistically nothing.
  The D10−D1 spread is positive in only **7 of 16** quarters — coin-flip consistency.
- **Weak hint, not a signal:** the *lowest*-consensus deciles (D1–D2, mostly single-holder stocks)
  underperform worst (~−9%), i.e. stocks nobody owns do slightly worse than the small-cap index;
  beyond that, more consensus buys you nothing measurable.

**Caveats (why this can't be pushed further):**
1. **Limited overlapping quarters:** deep panel coverage exists only for recent quarters
   (B2 backfill got static-HTML AMCs); usable (t, t+4q) windows number just 16, with per-quarter
   observation counts ranging from 6 to 352 — power is very low and windows overlap heavily.
2. **Benchmark mismatch:** the panel includes large/mid-cap names (HDFC Bank, Axis Bank…),
   while excess return is measured against NIFTY SMALLCAP 250 — which rallied ~50–65% through
   2023–24, mechanically pushing most stock-level excess returns negative.
3. **Consensus levels are low:** even D10 averages only ~4.8 holders, so "high consensus" here is
   modest in absolute terms.
4. **Price gaps:** `security_prices` has delisted-symbol gaps post-Jul-2024; exit dates use a
   ±45-day nearest-close tolerance around t+12 months (enforced ≥ ~4 forward quarters).

This is the correct honest outcome for a sanity backtest: the plumbing works; the naive
consensus→return hypothesis is **not supported** by the currently ingested data.

In [8]:
log.info("[DoD DEMO] one query: which small-cap schemes held %s, and what did it return vs the index?", DEMO_NAME)

# Part 1 — which small-cap schemes held the stock (all disclosures on record)
HOLDERS_SQL = f"""
SELECT DISTINCT a.name              AS amc,
       s.scheme_name                AS scheme,
       s.category                   AS category,
       ps.reporting_date            AS reporting_date,
       ph.percentage_to_nav::float8 AS pct_to_nav
FROM portfolio_holdings ph
JOIN portfolio_snapshots ps ON ps.id = ph.snapshot_id
JOIN schemes s ON s.id = ps.scheme_id
JOIN amcs a ON a.id = s.amc_id
WHERE ph.isin = '{DEMO_ISIN}'
  AND s.category ILIKE '%small cap%'
ORDER BY reporting_date DESC, amc, scheme
"""
holders = q(HOLDERS_SQL)
n_schemes = holders["scheme"].nunique() if len(holders) else 0
print(f"[DoD DEMO] {DEMO_NAME} ({DEMO_ISIN}): {len(holders)} small-cap-scheme holdings "
      f"across {n_schemes} schemes "
      f"(latest reporting_date: {holders['reporting_date'].max() if len(holders) else 'n/a'})")
print(holders.head(15).to_string(index=False))

# Part 2 — what the stock returned vs NIFTY SMALLCAP 250 over each t -> t+4q window it was held
PERF_SQL = f"""
WITH w AS (
    SELECT cp.qtr,
           e.close::float8 AS entry_close, e.trade_date AS entry_date,
           x.close::float8 AS exit_close,  x.trade_date AS exit_date
    FROM consensus_panel cp
    JOIN LATERAL (
        SELECT sp.close, sp.trade_date FROM security_prices sp
        WHERE sp.isin = '{DEMO_ISIN}' AND sp.trade_date <= cp.qtr
          AND sp.trade_date >= cp.qtr - INTERVAL '45 days'
        ORDER BY sp.trade_date DESC LIMIT 1
    ) e ON TRUE
    LEFT JOIN LATERAL (
        SELECT sp.close, sp.trade_date FROM security_prices sp
        WHERE sp.isin = '{DEMO_ISIN}'
          AND sp.trade_date >= cp.qtr + INTERVAL '12 months' - INTERVAL '45 days'
          AND sp.trade_date <= cp.qtr + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(sp.trade_date - (cp.qtr + INTERVAL '12 months')::date) LIMIT 1
    ) x ON TRUE
    WHERE cp.isin = '{DEMO_ISIN}'
),
perf AS (
    SELECT w.qtr::date AS qtr, w.entry_date, w.entry_close, w.exit_date, w.exit_close,
           (w.exit_close / w.entry_close - 1)::float8 AS stock_fwd_ret
    FROM w WHERE w.exit_close IS NOT NULL
)
SELECT p.qtr, p.entry_date, p.entry_close, p.exit_date, p.exit_close,
       round((p.stock_fwd_ret * 100)::numeric, 2)               AS stock_fwd_ret_pct,
       round((i.fwd_ret * 100)::numeric, 2)                     AS bench_fwd_ret_pct,
       round(((p.stock_fwd_ret - i.fwd_ret) * 100)::numeric, 2) AS excess_ret_pct
FROM perf p
JOIN LATERAL (
    SELECT (x.close / e.close - 1)::float8 AS fwd_ret
    FROM (SELECT 1) dummy
    JOIN LATERAL (
        SELECT ip.close FROM index_prices ip
        WHERE ip.index_symbol = '{BENCHMARK}' AND ip.trade_date <= p.qtr
        ORDER BY ip.trade_date DESC LIMIT 1
    ) e ON TRUE
    JOIN LATERAL (
        SELECT ip.close FROM index_prices ip
        WHERE ip.index_symbol = '{BENCHMARK}'
          AND ip.trade_date >= p.qtr + INTERVAL '12 months' - INTERVAL '45 days'
          AND ip.trade_date <= p.qtr + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(ip.trade_date - (p.qtr + INTERVAL '12 months')::date) LIMIT 1
    ) x ON TRUE
) i ON TRUE
ORDER BY p.qtr DESC
"""
perf = q(PERF_SQL)
print(f"\n[DoD DEMO] {DEMO_NAME} forward 12m return vs {BENCHMARK} by held quarter:")
print(perf.to_string(index=False))

2026-08-22 09:28:32,102 | INFO | [DoD DEMO] one query: which small-cap schemes held Sansera Engineering Limited, and what did it return vs the index?


[DoD DEMO] Sansera Engineering Limited (INE953O01021): 2 small-cap-scheme holdings across 2 schemes (latest reporting_date: 2026-07-31)
                           amc                            scheme                       category reporting_date  pct_to_nav
Baroda BNP Paribas Mutual Fund Baroda BNP Paribas Small Cap Fund Equity Scheme - Small Cap Fund     2026-07-31        1.95
               DSP Mutual Fund                DSP Small Cap Fund Equity Scheme - Small Cap Fund     2024-06-30        1.41

[DoD DEMO] Sansera Engineering Limited forward 12m return vs NIFTY SMALLCAP 250 by held quarter:
       qtr entry_date  entry_close  exit_date  exit_close  stock_fwd_ret_pct  bench_fwd_ret_pct  excess_ret_pct
2024-04-01 2024-04-01       1029.8 2025-04-01     1183.25               14.9               1.82           13.09
